lib/Sema/CSSimplify.cpp

```diff
@@ -7151,25 +7160,6 @@ ConstraintSystem::simplifyApplicableFnConstraint(
   // following: $T1 -> $T2.
   auto func1 = type1->castTo<FunctionType>();
 
-  // Let's check if this member couldn't be found and is fixed
-  // to exist based on its usage.
-  if (auto *memberTy = type2->getAs<TypeVariableType>()) {
-    if (isPotentialHole(memberTy)) {
-      auto *funcTy = type1->castTo<FunctionType>();
-      auto *locator = memberTy->getImpl().getLocator();
-      // Bind type variable associated with member to a type of argument
-      // application, which makes it seem like member exists with the
-      // types of the parameters matching argument types exactly.
-      addConstraint(ConstraintKind::Bind, memberTy, funcTy, locator);
```

Comment:
Yeah, I really like the idea of this, but I think it's happening in the wrong place, and doing it here loses information that the binding is for a hole, which is important. Ideally this could be done for all members, not just ones with an `applicable function` constraint. I'd like to experiment with seeing if we can do this in `getPotentialBindings` based on the constraints for a type variable.

Reply:
If we bound `memberTy` which is potentially a hole to a partially resolved type why would the fact that it could have been a hole be still important? By binding `memberTy` we know that it has a particular shape although that shape can have holes inside...

Reply:
It's important because we don't attempt bindings for a hole until there are no other bindings or disjunctions, so it can affect the order in which we bind things. Also, generic parameters are always potential holes – we don't actually know whether or not it's a hole until it's bound. In this case:
```swift
func g(f: (Int, Int) -> Int) {
  typealias Magic<T> = T
  _ = (f as Magic)(a: 0, b: 1)
}
```
We would end up binding the type variable for `T` to a function type in CSGen when adding the `applicable function` constraint, rather than inferring it normally while solving, and this changes the set of constraints at the end of CSGen – I don't think holes should affect that. This also seems to break some assumptions we've made when applying certain fixes (we don't produce the correct diagnostics for this test case and a few others). We could most likely change these assumptions to make it work, but I'm not convinced yet that it's necessary anyway, so I chose to remove it rather than trying to make it work. 

Regardless, I think it makes more sense to do this kind of thing in `getPotentialBindings` since we could do it more generally there.

Reply:
The problem in this example doesn't have anything to do with holes though right? it's related to `f` not marked as `@escaping` so binding `T` to `(Int, Int) -> Int` early seems okay... Also as far as remember this works - coercion is going to create a disjunction to convert `fn` to`Magic` which is going to be solved as part of a solver run anyway and not in CSGen?

Reply:
The other error is the extraneous argument labels `a:` and `b:` - this is what doesn't get diagnosed. You're right, there are no holes in this example, _but_ the code here executes anyway because generic parameters are always potential holes, and we don't know whether or not they are actually holes until they're bound. This code is checking for potential holes, so it doesn't know the difference

Reply:
What issue do you see with putting this code into `getPotentialBindings` to bind a hole to a more specific type if we can?

Reply:
bq. code here executes anyway and "solves" the applicable function constraint during CSGen because generic parameters are always potential holes

This actually shows a big limitation of a current model where we bind type variables to unresolved type to identify them as a hole. This block should be executed only if we know that type variable is a hole but at that point we can't re-assign the type variable to a different type since it's already bound :( Maybe a better solution would be to introduce a different flag - `TVO_IsHole` which is only set when we determine that there are no more bindings for a given type variable?

Reply:
That's exactly why I think the code for deciding which type to bind a hole to belongs in `getPotentialBindings`. Am I missing an issue with that approach?

Reply:
This example is not representative because `fn` already has a type, there are no holes there. 

Here is another one:

```swift
func foo() {
   return .bar(42)
}
```

I think member type for `bar` should be assumed as `(Int) -> Void`.

Reply:
Sure, I was just pointing out an example where this code causes problems. In the `.bar(42)` example, there is an applicable function constraint for what ends up being `(Int) -> Void` to the type variable of the unresolved member, so we could gather that information in `getPotentialBindings` when we iterate over the constraints.

Reply:
The approach I have described in the previous comment fits well for solving this. I don't think it's a job of potential bindings to be able to infer types for situations like that (ideally we wouldn't even need potential bindings in its current form because all of the possible types could be gathered from the constraint graph).